File: 0312-burst-balloons/solution.py
class Solution(object):
    def maxCoins(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums = [1] + nums + [1]
        
        n = len(nums)
        dp = [[0 for _ in range(n)] for __ in range(n)]
        
        # dp[i][i]   = 0
        # dp[i][i+1] = 0
        # dp[i][i+2] = nums[i]*nums[i+1]*nums[i+2]
        # dp[i][j] = max(dp[i][k]+dp[k+1][j] + nums[i]*nums[k]*nums[j]) for k [i+1, j]

        for length in range(3, n+2):
            for i in range(0, n - length+1):
                j = i + length - 1
                dp[i][j] = float('-inf')
                for k in range(i+1, j):
                    cost = nums[i]*nums[k]*nums[j]
                    dp[i][j] = max(dp[i][j], dp[i][k] + dp[k][j] + cost)
                    
        # for row in dp:
        #     print(row)
        
        return dp[0][n-1]
        

File: 0312-burst-balloons/solution.cpp
class Solution {
public:
    int maxCoins(vector<int>& nums) 
    {
        vector<int> matrices = {1};
        for(int i = 0; i < nums.size(); i++)
        {
            matrices.push_back(nums[i]);
        }
        matrices.push_back(1);
        
        const int n = matrices.size();

        vector<vector<int>> dp;
        for(int i = 0; i < n; i++)
        {
            dp.push_back(vector<int>(n-i, 0));
        }
        
        // for(auto & item: matrices)
        //         cout<<item<<" ";
        // cout<<endl;
    
        
        for(int length = 3; length <= n; length++)
        {
            for(int i = 0; i < n - length + 1; i++)
            {
                int j = i + length - 1;
                
                dp[i][j-i] = INT_MIN;
                
                for(int k = i+1; k < j; k++)
                {
                    int cost = matrices[i]*matrices[k]*matrices[j];
                    dp[i][j-i] = max(dp[i][j-i], dp[i][k-i] + dp[k][j-k] + cost);
                }
            }
        }
        
        // for(auto & row: dp)
        // {
        //     for(auto & item: row)
        //         cout<<item<<" ";
        //     cout<<endl;
        // }
        return dp[0][n-1];
        
        
        
    }
};
